# Lab | Pandas

### In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.


# Data Description

# 
. Customer - Customer ID

. ST - State where customers live

. Gender - Gender of the customer

. Education - Background education of customers

. Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

. Income - Customers income

. Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

. Number of Open Complaints - Number of complaints the customer opened

. Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

. Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

. Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.

External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9



# Challenge 1: Understanding the data
In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

. Identify the dimensions of the dataset by determining the number of rows and columns it contains.

In [2]:
import numpy as np
import pandas as pd

In [3]:
# Load Customer Data from an insurance company from an online source
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'

# read_csv
df = pd.read_csv(url)
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [4]:
#. Identify the dimensions of the dataset by determining the number of rows and columns it contains.
shape = df.shape
df.shape
print('\ncustomer data shape:', shape)


customer data shape: (4008, 11)


In [5]:
tuple(customer_data_df.shape)

NameError: name 'customer_data_df' is not defined

#
Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable.
#You should also provide suggestions for fixing any incorrect data types

In [ ]:
# to get the datatypes of each column
df.dtypes

In [ ]:
#Following suggestions for the datatypes 
# 1.we can change the datatypes for "Gender" as "Categorical type" as it can have fixed values , 
# 2. 'No of open complaints" can be "numerical type" instead of object as the values will be numerical only
# 3. "Policy type" as 'categorical type'.As I can see it has fixed values.
# 4. And customer lifetime value is in % so we can multiply it with 100 and change the dtype to numerical type.

#
Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.


In [6]:
# Identify number of unique values for each column
##num_unique_values = df.nunique()

num_unique_values = df.nunique()

print("Number of Unique Values for Each Column:")
print(num_unique_values)
print()


Number of Unique Values for Each Column:
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64



In [7]:
# Determine categorical columns:
#categorical_columns = df.select_dtypes(include=["object"]).columns
categorical_columns = df.select_dtypes(include=["object"]).columns
print("Categorical Columns:")
print(categorical_columns)
display()



Categorical Columns:
Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Number of Open Complaints', 'Policy Type', 'Vehicle Class'],
      dtype='object')


In [8]:
# Describe unique values of categorical columns
for column in categorical_columns:
    #unique_values = df[column].unique()   #(for my reference)
    unique_values = df[column].unique()
    print(f"Unique values in '{column}': {unique_values}")
display()

Unique values in 'Customer': ['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]
Unique values in 'ST': ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
Unique values in 'GENDER': [nan 'F' 'M' 'Femal' 'Male' 'female']
Unique values in 'Education': ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
Unique values in 'Customer Lifetime Value': [nan '697953.59%' '1288743.17%' ... '2031499.76%' '323912.47%'
 '899704.02%']
Unique values in 'Number of Open Complaints': ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]
Unique values in 'Policy Type': ['Personal Auto' 'Corporate Auto' 'Special Auto' nan]
Unique values in 'Vehicle Class': ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


In [9]:
# Describe numerical columns
#numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns   #(for my reference)
numerical_columns = df.select_dtypes(include='number').columns 
print("Numerical Columns:")
print(numerical_columns)
print()

Numerical Columns:
Index(['Income', 'Monthly Premium Auto', 'Total Claim Amount'], dtype='object')



In [10]:
#numerical_description = df[numerical_columns].describe()
numerical_description = df[numerical_columns].describe()

print("Description of Numerical Columns:")
print(numerical_description)
print()

Description of Numerical Columns:
             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678



In [11]:
# Insights:
#1.we can see the data distribution.
#2.We can check the unique values for categorical columns and for numerical columns, we can check it with a 
# description method which will include statistics like count, mean, standard deviation, minimum, maximum, and quartiles.
#3.By identifying categorical and numerical columns it provides us an overview of our dataset characteristics.


# 
 #Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

In [12]:
# Compute summary statistics for numerical columns

#numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns   (for my reference)
numerical_columns = df.select_dtypes(include='number').columns 

# .describe() method is used to compute summary statistics for these numerical columns.
summary_statistics = df[numerical_columns].describe()

print("Summary Statistics for Numerical Columns:")
display(summary_statistics)

Summary Statistics for Numerical Columns:


Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

# 
From the data given by Summary statistics,following conclusions can be provided for each column:
1. Income : The mean 'Income' is around 39295.70, the median (50th percentile) is 36234.00
            The standard deviation is approx 30469, 
            The quartiles indicate that the range is from 14072 to 99960 
2. Monthly Premium Auto: 
3. Total Claim Amount:
    note: I have not done with other columns because I am not sure whether the way I am doing is right.

# 
Summary statistics provide a quick overview of the data distribution and central tendency.
It can help us to identify outliers, understand the spread of data, and make initial observations about
the dataset's characteristics.

# 
#Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [13]:
# Compute summary statistics for categorical columns

#categorical_columns = df.select_dtypes(include=["object"]).columns    #(For my reference)

categorical_columns = df.select_dtypes(include=["object"]).columns

categorical_statistics = df[categorical_columns].describe()

print("Summary Statistics for Categorical Columns:")
display(categorical_statistics)


Summary Statistics for Categorical Columns:


Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

# Conclusions: 
1.The most frequent Gender is "Female" having a count of 954

2. The most frequent vehicle class is "576" having a count of 1071

3. The top most used Education Type is "Bachelor" having count of '1071'

4. Since 'No of open Complaints' are '1071' but unique given is '6' so we have to look into 6 types of complaints

# Challenge 2: analyzing the data

# Exercise 1
The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [19]:
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [15]:
# To get the customer location
customer_location = df.ST
display(df.ST)

0       Washington
1          Arizona
2           Nevada
3       California
4       Washington
           ...    
4003           NaN
4004           NaN
4005           NaN
4006           NaN
4007           NaN
Name: ST, Length: 4008, dtype: object

In [23]:

# Calculate the frequency of each customer location
location_counts = df['ST'].value_counts()


In [24]:
# Sort the counts in ascending order
location_counts_sorted = location_counts.sort_values()

In [25]:
# Retrieve the top 5 less common locations
top_5_less_common_locations = location_counts_sorted.head(5)

print("Top 5 Less Common Customer Locations (Ascending Order):")
print(top_5_less_common_locations)

Top 5 Less Common Customer Locations (Ascending Order):
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: ST, dtype: int64


# Exercise 2
The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

Hint:

Using value_counts() method simplifies this analysis.
Futhermore, there is a method that returns the index of the maximum value in a column or row.

In [27]:

# Count the number of policies sold for each policy type
policy_counts = df['Policy Type'].value_counts()

# Retrieve the policy type with the highest number of policies sold
most_sold_policy = policy_counts.idxmax()

print("Total number of policies sold for each type:")
print(policy_counts)
print("\nPolicy type with the highest number of policies sold:", most_sold_policy)


Total number of policies sold for each type:
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: Policy Type, dtype: int64

Policy type with the highest number of policies sold: Personal Auto


# Exercise 3
The sales team wants to know if customers with Personal Auto have a income than those with Corporate Auto. How does the average income compare between the two policy types?

Use loc to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
Calculate the average income for each policy.
Print the results.

In [29]:


# Create a DataFrame for Personal Auto policies
personal_auto_df = df.loc[df['Policy Type'] == 'Personal Auto']

# Create a DataFrame for Corporate Auto policies
corporate_auto_df = df.loc[df['Policy Type'] == 'Corporate Auto']

# Calculate the average income for each policy type
average_income_personal_auto = personal_auto_df['Income'].mean()
average_income_corporate_auto = corporate_auto_df['Income'].mean()

print("Average Income for Personal Auto Policies:", average_income_personal_auto)
print("Average Income for Corporate Auto Policies:", average_income_corporate_auto)


Average Income for Personal Auto Policies: 38180.69871794872
Average Income for Corporate Auto Policies: 41390.31196581197


# Bonus: Exercise 4
Your goal is to identify customers with a high policy claim amount.

Instructions:

Review again the statistics for total claim amount to gain an understanding of the data.
To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
Use DataFrame methods to calculate summary statistics about the high policy claim amount data.
Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.

Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.

Hint 2: check Boolean selection according to the values of a single column in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9